### DISCUSSION TITLE
Folium's future

### DISCUSSION THREAD

**Discussion by ocefpaf** [python-visualization/folium#1955] [Collaborators]
@Conengmo and @hansthen, I've been thinking about Folium's future lately and I'd like to pass two ideas for your consideration:

1. My day-job org is participating in Google Summer of Code (GSoC) for the 3rd time and, hopefully, we will be selected again for 2025. I'd love to have an ideas page for Folium so we can try to get more people involved. What do you think? Any ideas that we could use? Would any of you be available to act as mentors?
2. I wonder if we should apply to be a [NumFOCUS (NF) affiliated project](https://numfocus.org/sponsored-projects/affiliated-projects). That would give us access to another GSoC umbrella, the possibility to apply for Small Development Grants, more recognition and a chance to grow the community.

I can help with 1 if you agree but I cannot do 2 b/c I am part of the NF committee that evaluates the applications. I would abstain from applying and reviewing Folium's application.

**Comment by hansthen** [comment]
1. What a good idea! Would the ideas page be separate from the github issues list? I have a few ideas to put on the ideas page, which I think are suitable for GSoC. About the mentorship, what time involvement would be required for it?

2. My first reaction was that Folium is less numerically focused than the other projects. This is not saying we should not do it, but I do not see a straightforward path for an application. I will check the NumFocus web page better later.

**Reply by ocefpaf** [reply]
For GSoC we can start with an issue, or labeling issues. When the time comes we would send a compilation to https://github.com/ioos/gsoc.

The mentorship commitment changed a lot in the latest editions of GSoC, they now have various projects sizes. We would need to find an optimum for the time we have. For 2024 they were: Small: ~90 hours. Medium: ~175 hours. Large: ~350 hours.

Regarding NumFOCUS, the Num in the name is misleading, any project with scientific orientation is a valid candidate. Maps definitely have that use case (it is my use case).

**Reply by Conengmo** [reply]
An Ideas page would also fit very well in the Discussions thing we have now!

**Reply by hansthen** [reply]
> The mentorship commitment changed a lot in the latest editions of GSoC, they now have various projects sizes. We would need to find an optimum for the time we have. For 2024 they were: Small: ~90 hours. Medium: ~175 hours. Large: ~350 hours.

Over what period would that commitment be spread? I may be able to commit to mentoring a small project.

**Comment by Conengmo** [comment]
Thanks for bringing these ideas Filipe!

1. I'm afraid I don't really have time to help mentor over the summer. I'm 'behind' Folium work now already, so I don't want to take on additional responsibility. I'm also not sure what would be good issues for this, since the low hanging fruit has been long picked, and most of the issues or enhancements we have maybe require more experience with the project. But that sounds a bit old and grumpy doesn't it :)
2. Love the idea. I'm very interested in making sure Folium stays working in the long term, and this might help? Having a community that we can contact for help when we need it. Or possibly get input in our community on how we can be better right now. Next application round starts July 15th. We should have a Code of Conduct before we apply. I understand you can't apply, I'm motivated to look into this.

While we're talking in a thread called 'Folium's future', maybe we can also discuss how we see Folium's role in the coming years. I have thoughts about this, but I also love to hear how you all see this.

In my view, Folium, with all it's flaws and idiosyncrasies :), is pretty stable and mature. It has been around for quite some time. It wraps Leaflet.js, which itself is stable and mature. (Leaflet [talks of a 2.0 version](https://leafletjs.com/2022/09/21/leaflet-1.9.0.html) at some point in the future, which will be mainly about better performance under the hood.)

Folium is featured in many tutorials on the internet, and I suspect also in courses in education. It's also used by other packages downstream, that have gotten popular themselves.

So at this point I see it as my responsibility to make sure Folium stays working for those users, and to me that is more important than new features or optimizations under the hood.

That's the demand side. On the supply side we have a small team of core people with limited time, so our capabilities are limited. Drive-by users are helpful, but we don't have an appealing backlog for them to work on, so they don't generally stay.

In conclusion, I see Folium as being in a sort of maintenance mode. Because others depend on it, and we don't have much manpower to work on it.

If new technology or new ideas come up on how creating maps in Python could be done better, I'd be hesitant to work that into Folium, risking breaking it. Stuff like better interactivity came and has successfully been done in other projects, some of them even relying on Folium. Next, maybe Javascript/Python interaction will become better with new libraries, or some other thing we don't yet know about. I'd argue to keep Folium what it is now, a static Leaflet.js generator for Python, and do rewrites or new approaches in other projects.

So those are my thoughts, I'd love to hear how you see this, what Folium should stay or should become, and how we should work on it.

**Reply by hansthen** [reply]
I agree with almost everything. I see Leaflet as stable and mature. We should optimize Folium for maintenance. I also think Folium should stay a static Leaflet.js generator for Python. (Although I mostly use Folium via streamlit-folium and I am very impressed with how they added interactivity to Folium.) 

I am a bit hesitant to speak up since I have only recently joined, but I have a different perspective on a few minor points. 

1. When I started Folium there were quite a few edge cases where things did not work as I expected them to work. A few examples:

a. Sometimes you need to `add` stuff to a map where this is not required in `Leaflet`. @Conengmo you also noted this a few times in my Pull Requests. 
b. Folium has a `DivIcon`, but you cannot always use this in place of an `Icon`. You cannot use it inside a `GeoJson` layer for instance.
c. You cannot re-use icons over multiple layers.

2. I also found it difficult to understand the class structure and the relationship between Folium and Branca. 

As I see it now, the first is related to the second. The way I understand Branca, is that it is a template modelling library modeled after the structure of HTML. The basic metaphor is true composition. Everything is related via parent-child relationships. The parent really is the sum of its children. Children are part of only one parent.

This metaphor (composition) can *sometimes* be applied to `Leaflet`, but not *always*. `L.Map` and `L.FeatureGroup` can function as parents in a composition relationship, but many other items cannot. Also, Leaflet can use pass by reference for its elements, which makes that some children can have multiple parents.

3. Many plugins feel like they are written over a time. Code sometimes feels copied and pasted and then adapted. Also coding styles seem to have evolved during that time.

4. Folium is missing core functionality from Leaflet that makes it more difficult to extend Folium.

These four items add to the maintenance burden and also make up quite a few items on the issues list. What I would like to do, if you let me, is rewrite some things in Folium. This is meant to reduce the maintenance burden in the long run. I use Folium extensively at work and I can use (some of) my paid hours to work on Folium.

Item 3 requires good old cleanup work. For items 1, 2 and 4 I have some ideas. They will involve rewrites, but they are comparatively minor and not a completely new approach. They can also be implemented largely maintaining backward compatibility. Users should not have to be educated about the differences. I also think I can make the changes incrementally and discuss them with you before I go ahead with each step.

**Reply by ocefpaf** [reply]
I'll comment more on all the points later but this is more important.

> I am a bit hesitant to speak up since I have only recently joined

Don't be. You are a maintainer now.

> but I have a different perspective on a few minor points.

That is very welcomed and we can certainly reconcile our views and find a path forward.

**Reply by Conengmo** [reply]
Totally agree with Filipe here. I’ve been looking at this project for some years now so I’m a bit stuck in my thinking.

**Comment by ocefpaf** [comment]
> In conclusion, I see Folium as being in a sort of maintenance mode. Because others depend on it, and we don't have much manpower to work on it.

I agree. In fact, I used to think that before you steeped in! We sure can add small improvements here and there but I'd love to keep the core as stable as possible and running with little burden maintenance on you two.

Regarding 1, got it, without mentor we can't do it.
For 2, it is a 1-form application and then, if approved, we have access to some perks that we don't necessarily need to use if we don't have time to deal with it.

**Reply by Conengmo** [reply]
About 2, sounds good! I'll add a Code of Conduct then look into applying afterwards. @hansthen, do you think it's worth applying as well? Folium is scientific oriented, so in terms of orientation it's probably a good fit.

**Reply by hansthen** [reply]
Yes, I think we can apply.

**Reply by ocefpaf** [reply]
BTW, next release, folium 1.0? 😬

**Comment by Conengmo** [comment]
@ocefpaf in the Numfocus application they ask to list all the people who have write access to the repo. Maybe this is a good moment to go through https://github.com/orgs/python-visualization/people and make anyone not currently actively helping out an honorary member (without write access). Is that okay to you?

**Reply by ocefpaf** [reply]
Since the xz incident I have been thinking about that. I was always very eager to get people commit rights and let them feel owners/peers of folium and folium devs. However, if a person is MIA and their credentials get stolen, we can't do much.

TL;DR I'm OK and, if someone contacts us b/c they want to be back onboard, we can ask them for a plan on their future contributions, and restore their admin rights.

**Reply by Conengmo** [reply]
Sounds good! Agree with the dilemma of welcoming people vs. security, that's a difficult one.

**Reply by Conengmo** [reply]
I cleaned up some access rights here and there. Some more thoughts/questions:

- can we de-own wrobstory? He's the original creator, so it seems weird to take away his rights. On the other hand, any unused rights are a possible security risk...

**Reply by ocefpaf** [reply]
We should probably ask him. Although it is super hard to get a hold of him lately. Maybe a gentle ping in a clean issue/message in the discussion board? Email would be better but I don't think I have one.

**Reply by Conengmo** [reply]
Good call. I have his email, I'll drop him an email asking what he would prefer.

**Comment by Conengmo** [comment]
I did the NumFOCUS application! We'll see what happens next...

**Comment by Conengmo** [comment]
We got accepted as affiliate at NumFOCUS! I'll share your contact details with them, so we can get added to their communication channels.
